File: src/net/presence.ts
/* How many people are looking at the canvas.
 *
 * The archive holds the count, because it can hold an exact last-seen time and expire it.
 * technocore.chat's presence convention cannot: its notes have no expiry, `GET /kv/<ns>` lists
 * key names without a write time, and the value it standardises is "the seq you last saw" —
 * which stops distinguishing a live viewer from an old one exactly when the canvas is quiet.
 * Using it would also leave one permanent note per visitor in a shared public namespace.
 *
 * The id is invented per page load and is not the visitor's did:key. A viewer count has no
 * business learning which identity is reading, and a per-load id counts what the number
 * claims to count: open tabs, not people and not keys.
 *
 * A hidden tab does not beat. It is not looking, and a browser suspends its timers anyway, so
 * beating on the way out would only mean claiming a viewer that is not there for as long as
 * the window lasts.
 */

const BEAT_MS = 15_000;
const TIMEOUT_MS = 8_000;

/** Told to the caller when the count is unknown, so it can say nothing rather than "0". */
export type Watching = number | null;

export interface PresenceEvents {
  /** The count changed, or became unknown. `capped` means the answer is a floor. */
  onCount(watching: Watching, capped: boolean): void;
}

const viewerId = (): string => {
  const bytes = new Uint8Array(8);
  crypto.getRandomValues(bytes);
  return [...bytes].map((b) => b.toString(16).padStart(2, "0")).join("");
};

/**
 * Beat while this tab is visible, reporting the count each time.
 *
 * Returns a function that stops it. Never throws: a viewer count that could break the page it
 * decorates would be worse than no viewer count, so a failed beat reports "unknown" and the
 * next one tries again.
 */
export function follow(archiveUrl: string, events: PresenceEvents): () => void {
  const id = viewerId();
  let timer = 0;
  let stopped = false;
  let inFlight: AbortController | null = null;

  const beat = async (): Promise<void> => {
    if (stopped || document.hidden) return;
    inFlight?.abort();
    const abort = new AbortController();
    inFlight = abort;
    const bail = setTimeout(() => abort.abort(), TIMEOUT_MS);
    try {
      const response = await fetch(`${archiveUrl}/presence/${id}`, { signal: abort.signal });
      if (!response.ok) throw new Error(String(response.status));
      const body: unknown = await response.json();
      const viewers = (body as { viewers?: unknown }).viewers;
      const capped = (body as { capped?: unknown }).capped === true;
      // Checked rather than trusted: this is drawn into the page, and a number that arrived
      // as a string would render as one.
      if (typeof viewers !== "number" || !Number.isFinite(viewers) || viewers < 0) {
        throw new Error("the archive did not answer with a count");
      }
      if (!stopped) events.onCount(Math.floor(viewers), capped);
    } catch {
      // Includes the abort on stop, which is why nothing is reported after stopping.
      if (!stopped) events.onCount(null, false);
    } finally {
      clearTimeout(bail);
      if (inFlight === abort) inFlight = null;
    }
  };

  const onVisibility = (): void => {
    // Coming back should show a current number rather than the one from before the tab was
    // hidden, which by now is as old as the absence.
    if (!document.hidden) void beat();
  };

  void beat();
  timer = window.setInterval(() => void beat(), BEAT_MS);
  document.addEventListener("visibilitychange", onVisibility);

  return () => {
    stopped = true;
    window.clearInterval(timer);
    document.removeEventListener("visibilitychange", onVisibility);
    inFlight?.abort();
  };
}

File: server/canvas/presence.py
"""Who is looking at the canvas right now.

This is the one number on the page that cannot be proved, and the interface says so. Every
other figure rests on a signature: a pixel is witnessed because the signature verified here.
A viewer count rests on browsers volunteering that they exist, and a browser can say anything.
It is a courtesy figure, not evidence, and dressing it up as evidence would undo the one claim
this product actually makes.

Why it is not technocore.chat's presence convention (`/kv/<room>/hb-<nick>/set/<seq>`),
checked against the live service on 2026-09-03:

  - notes have no server-side expiry, so one key per visitor accumulates forever in a shared
    public namespace capped at 131072 — a leak into someone else's service, not just ours;
  - `GET /kv/<ns>` lists key names only, with no write time, so liveness has to be inferred
    from the value, which is "the seq you last saw";
  - that inference fails exactly when this canvas is quiet. With no placements for a day,
    every heartbeat still holds the same seq, so someone who visited last week reads as
    present. A count that cannot tell a live viewer from an old one is a count that lies.

So presence is held here, in memory, where a last-seen time is exact and expiry is real.
Nothing is persisted: this answers "now", and a restart correctly forgets everyone.
"""

from __future__ import annotations

import re
import threading

# 16 hex characters, generated per page load in the browser. Not the visitor's did:key: a
# viewer count has no business learning which identity is reading, and a per-load id counts
# what the number claims to count — open tabs — rather than people or keys.
VIEWER_PATTERN = re.compile(r"^[0-9a-f]{16}$")

# Long enough that one missed beat does not drop a viewer, short enough that a closed tab
# leaves promptly. The client beats at a third of this.
DEFAULT_WINDOW_SECONDS = 45.0

# A ceiling on what one process will track. Reached only under a flood of invented ids, which
# nothing can prevent — see the module docstring.
DEFAULT_CAPACITY = 20_000


class Presence:
    """Last-seen times for viewers, bounded in size and swept on every use.

    Safe to call from several request threads: the server hands each connection its own
    thread, so an unguarded dict would drop beats under concurrent writes.
    """

    def __init__(
        self,
        window_seconds: float = DEFAULT_WINDOW_SECONDS,
        capacity: int = DEFAULT_CAPACITY,
    ) -> None:
        if window_seconds <= 0:
            raise ValueError("window_seconds must be positive")
        if capacity <= 0:
            raise ValueError("capacity must be positive")
        self._window = window_seconds
        self._capacity = capacity
        self._seen: dict[str, float] = {}
        self._lock = threading.Lock()

    @property
    def capacity(self) -> int:
        return self._capacity

    def beat(self, viewer: str, now: float) -> tuple[int, bool]:
        """Record that `viewer` is here, and answer how many are.

        Returns the count and whether the tracker is at capacity, so the caller can say
        "20000+" rather than state a number it knows is a floor.

        :raises ValueError: if `viewer` is not 16 hex characters.
        """
        if not VIEWER_PATTERN.match(viewer):
            raise ValueError("viewer must be 16 hex characters")
        with self._lock:
            self._sweep(now)
            if viewer not in self._seen and len(self._seen) >= self._capacity:
                return len(self._seen), True
            self._seen[viewer] = now
            return len(self._seen), len(self._seen) >= self._capacity

    def count(self, now: float) -> int:
        """How many viewers have beaten within the window."""
        with self._lock:
            self._sweep(now)
            return len(self._seen)

    def _sweep(self, now: float) -> None:
        """Drop viewers whose last beat fell out of the window. Caller holds the lock.

        A clock that moves backwards would otherwise strand entries forever, so a last-seen
        time in the future is treated as now rather than trusted.
        """
        cutoff = now - self._window
        stale = [viewer for viewer, at in self._seen.items() if at <= cutoff]
        for viewer in stale:
            del self._seen[viewer]
